### PR TITLE
Call api.HashLookup only once for a batch query

### PR DESCRIPTION
**Pull request by panqingfeng**
Call api.HashLookup only once for a batch query

### CODE REVIEW

File: safebrowser.go [hunk 1]
```diff
@@ -338,6 +340,8 @@ func NewSafeBrowser(conf Config) (*SafeBrowser, error) {
 		}
 	}
 
+	sb.nextlookuptime = time.Unix(0,0)
```
Comment (colonelxc) [inline safebrowser.go]: How about let's skip the MinWaitDuration pieces in this change, and do it in a separate pull request.

File: safebrowser.go [hunk 2]
```diff
@@ -405,30 +409,34 @@ func (sb *SafeBrowser) LookupURLs(urls []string) (threats [][]URLThreat, err err
 	// TODO: There are some optimizations to be made here:
 	//	1.) We could force a database update if it is in error.
 	//  However, we must ensure that we perform some form of rate-limiting.
-	//	2.) We should batch all of the partial hashes together such that we
-	//	call api.HashLookup only once.
+	hashes := make(map[hashPrefix]string)
```
Comment (colonelxc) [inline safebrowser.go]: hash2pattern? Mirrors your hash2idx below.

File: safebrowser.go [hunk 3]
```diff
@@ -405,30 +409,34 @@ func (sb *SafeBrowser) LookupURLs(urls []string) (threats [][]URLThreat, err err
 	// TODO: There are some optimizations to be made here:
```
Comment (colonelxc) [inline safebrowser.go]: I think the whole TODO could be deleted now. Along with the one you are fixing here, the DB now does proper retrying according to the specified backoff schedule. So there's not really a situation in which we should trigger an update while doing a lookup. On receiving a failure indicating a DB issue, the caller could call WaitUntilReady(ctx) and get the right behavior (Ideally we will have public errors that make it easy to tell if it is a DB issue versus other types of errors).

File: safebrowser.go [hunk 4]
```diff
@@ -463,34 +472,39 @@ func (sb *SafeBrowser) LookupURLs(urls []string) (threats [][]URLThreat, err err
...
-			atomic.AddInt64(&sb.stats.QueriesFail, int64(len(urls)-i))
+			atomic.AddInt64(&sb.stats.QueriesFail, 1)
 			return threats, err
 		}
 
+		if resp.MinimumWaitDuration != nil {
+			serverMinWait := time.Duration(resp.MinimumWaitDuration.Seconds)*time.Second + time.Duration(resp.MinimumWaitDuration.Nanos)
+                        sb.log.Printf("Next lookup in %v", serverMinWait)
+			sb.nextlookuptime = time.Now().Add(serverMinWait)
+                } else {
```
Comment (colonelxc) [inline safebrowser.go]: Looks like a formatting error here. Please run gofmt.

File: safebrowser.go (hunk 1 above)
Comment (panqingfeng) [inline safebrowser.go]: mmm...I am not sure what "separate pull request" mean, should I discard this "MinWaitDuration“ commit?

I am a c/c++ engineer and new to go, I am working on a email antispam project that scan 10M emails every day, so I have to sent as less request as possible. Base on the doc of Google, a client should respect server's MinWaitDuration respond, that's why I make the change. Can "a separate pull request" work it around?
Thank you again

File: safebrowser.go (hunk 3 above)
Comment (panqingfeng) [inline safebrowser.go]: OK, that TODO are removed.

File: safebrowser.go (hunk 4 above)
Comment (panqingfeng) [inline safebrowser.go]: It's done, thanks for the advice

File: safebrowser.go (hunk 2 above)
Comment (panqingfeng) [inline safebrowser.go]: I don't follow on this either. Yes, this hashes is hash2pattern. but hash2idx store the index of "threats [][]URLThreat".

Comment (colonelxc) [inline safebrowser.go]: I was just suggesting a variable name change, but leaving it the same is fine.

File: safebrowser.go (hunk 1 above)
Comment (colonelxc) [inline safebrowser.go]: I was trying to say that we should discard the MinWaitDuration commit for now, and then once the batching is merged, send the MinWaitDuration commit for review separately. Do that and I can merge this now.
